ktor-client/ktor-client-darwin/darwin/src/io/ktor/client/engine/darwin/DarwinWebsocketDelegate.kt

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright 2014-2022 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.client.engine.darwin
+
+import io.ktor.client.request.*
+import io.ktor.http.*
+import io.ktor.util.date.*
+import io.ktor.utils.io.core.*
+import io.ktor.websocket.*
+import kotlinx.cinterop.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.*
+import platform.Foundation.*
+import platform.darwin.*
+import kotlin.coroutines.*
+
+@OptIn(UnsafeNumber::class, DelicateCoroutinesApi::class, ExperimentalCoroutinesApi::class)
+internal class DarwinWebsocketSession(
+    private val requestTime: GMTDate,
+    coroutineContext: CoroutineContext
+) : WebSocketSession {
+
+    internal lateinit var task: NSURLSessionWebSocketTask
+
+    internal val response = CompletableDeferred<HttpResponseData>()
+
+    private val _incoming = Channel<Frame>(Channel.UNLIMITED)
+    private val _outgoing = Channel<Frame>(Channel.UNLIMITED)
+    private val socketJob = Job(coroutineContext[Job])
+    private val receiveJob = Job(socketJob)
+    override val coroutineContext: CoroutineContext = coroutineContext + socketJob
+
+    override var masking: Boolean
+        get() = true
+        set(_) {}
+
+    override var maxFrameSize: Long
+        get() = task.maximumMessageSize.convert()
+        set(value) {
+            task.setMaximumMessageSize(value.convert())
+        }
+
+    override val incoming: ReceiveChannel<Frame>
+        get() = _incoming
+
+    override val outgoing: SendChannel<Frame>
+        get() = _outgoing
+
+    override val extensions: List<WebSocketExtension<*>>
+        get() = emptyList()
+
+    public val delegate: NSURLSessionWebSocketDelegateProtocol = WebSocketDelegate()
+
+    init {
+        launch {
+            receiveMessages()
+        }
+        launch {
+            sendMessages()
+        }
+        GlobalScope.launch(coroutineContext) {
+            try {
+                this.coroutineContext[Job]!!.join()
+            } catch (cause: Exception) {
+                val code = CloseReason.Codes.INTERNAL_ERROR.code.convert<NSInteger>()
+                task.cancelWithCloseCode(code, "Client failed".toByteArray().toNSData())
+            } finally {
+                _incoming.close()
+                _outgoing.cancel()
+            }
+        }
+    }
+
+    private suspend fun receiveMessages() {
+        val message = withContext(receiveJob) {
```

Comment:
why do we need withContext here?

Reply:
To be able to cancel receiving of the message. Otherwise `task.receiveMessageWithCompletionHandler` hangs

Reply:
But, it's strange, as `receiveMessages` is already called inside `launch` with parent Job of `socketJob`, same as with `receiveJob` so it should have the same job hierarchy
May be it's possible to remove it, but during `init` assign Job from `launch` call to `receiveJob` and use it for cancellation of receiving messages?

Reply:
`socketJob` can't wake up and cancel `receiveMessageWithCompletionHandler` when you call `complete` on it. Only `completeExceptionally` does it. 
Using `launch` to create an `receiveJob` doesn't work, because `launch` returns `Job` and not `CompletableJob`.

Reply:
You are right, `cancel` did the trick